File: nexus-adapter-sdk-go/stream.go
package nexadapter

import (
	"bufio"
	"context"
	"encoding/json"
	"fmt"
	"os"
)

// StreamConfig configures the streaming delivery handler.
// Only needed for adapters that declare "stream" in their supports.
type StreamConfig struct {
	// OnStreamStart is called when a new streaming delivery begins.
	// The adapter should create a platform message and prepare for token accumulation.
	OnStreamStart func(ctx context.Context, event StreamEvent) error

	// OnToken is called for each token in the stream.
	// The adapter should buffer tokens and periodically update the platform message
	// (e.g., Discord edits every ~300ms, Telegram editMessageText).
	OnToken func(ctx context.Context, event StreamEvent) error

	// OnToolStatus is called when a tool execution status changes.
	// The adapter can optionally display tool activity indicators.
	OnToolStatus func(ctx context.Context, event StreamEvent) error

	// OnReasoning is called for reasoning/thinking tokens.
	// The adapter can optionally display these (e.g., in a spoiler block).
	OnReasoning func(ctx context.Context, event StreamEvent) error

	// OnStreamEnd is called when the streaming delivery is complete.
	// The adapter should finalize the message and report delivery status.
	OnStreamEnd func(ctx context.Context, event StreamEvent) error

	// OnStreamError is called when an error occurs during streaming.
	// The adapter should handle partial delivery gracefully.
	OnStreamError func(ctx context.Context, event StreamEvent) error
}

// EmitStreamStatus writes an AdapterStreamStatus to stdout.
// Call this from stream callbacks to report delivery progress to NEX.
func EmitStreamStatus(status AdapterStreamStatus) error {
	return writeJSON(status)
}

// handleStream reads StreamEvent JSONL from stdin and dispatches to callbacks.
// This is the main loop for the `stream` command.
func handleStream(ctx context.Context, config *StreamConfig) error {
	scanner := bufio.NewScanner(os.Stdin)
	// Increase buffer size for potentially large events
	scanner.Buffer(make([]byte, 0, 64*1024), 1024*1024)

	for scanner.Scan() {
		select {
		case <-ctx.Done():
			LogInfo("stream shutting down (context cancelled)")
			return nil
		default:
		}

		line := scanner.Bytes()
		if len(line) == 0 {
			continue
		}

		var event StreamEvent
		if err := json.Unmarshal(line, &event); err != nil {
			LogError("stream: failed to parse event: %v", err)
			continue
		}

		var err error
		switch event.Type {
		case "stream_start":
			if config.OnStreamStart != nil {
				err = config.OnStreamStart(ctx, event)
			}
		case "token":
			if config.OnToken != nil {
				err = config.OnToken(ctx, event)
			}
		case "tool_status":
			if config.OnToolStatus != nil {
				err = config.OnToolStatus(ctx, event)
			}
		case "reasoning":
			if config.OnReasoning != nil {
				err = config.OnReasoning(ctx, event)
			}
		case "stream_end":
			if config.OnStreamEnd != nil {
				err = config.OnStreamEnd(ctx, event)
			}
		case "stream_error":
			if config.OnStreamError != nil {
				err = config.OnStreamError(ctx, event)
			}
		default:
			LogDebug("stream: unknown event type: %s", event.Type)
		}

		if err != nil {
			LogError("stream handler error for %s: %v", event.Type, err)
			_ = EmitStreamStatus(AdapterStreamStatus{
				Type:     "delivery_error",
				ErrorMsg: err.Error(),
			})
		}
	}

	if err := scanner.Err(); err != nil {
		return fmt.Errorf("stream: stdin read error: %w", err)
	}

	return nil
}

File: nexus-adapter-sdks/nexus-adapter-sdk-ts/src/stream.ts
import { AdapterLogger } from "./logger.js";
import { readJSONLines, writeJSONLine } from "./io.js";
import {
  AdapterStreamStatusSchema,
  StreamEventSchema,
  type AdapterStreamStatus,
  type StreamEvent,
} from "./protocol.js";
import type { AdapterContext } from "./run.js";

export type StreamHandlers = {
  onStart?: (ctx: AdapterContext, event: Extract<StreamEvent, { type: "stream_start" }>) => void | Promise<void>;
  onToken?: (ctx: AdapterContext, event: Extract<StreamEvent, { type: "token" }>) => void | Promise<void>;
  onToolStatus?: (
    ctx: AdapterContext,
    event: Extract<StreamEvent, { type: "tool_status" }>,
  ) => void | Promise<void>;
  onReasoning?: (
    ctx: AdapterContext,
    event: Extract<StreamEvent, { type: "reasoning" }>,
  ) => void | Promise<void>;
  onEnd?: (ctx: AdapterContext, event: Extract<StreamEvent, { type: "stream_end" }>) => void | Promise<void>;
  onError?: (
    ctx: AdapterContext,
    event: Extract<StreamEvent, { type: "stream_error" }>,
  ) => void | Promise<void>;
};

export function emitStreamStatus(
  status: AdapterStreamStatus,
  stdout: NodeJS.WriteStream = process.stdout,
  validate = true,
): void {
  const payload = validate ? AdapterStreamStatusSchema.parse(status) : status;
  writeJSONLine(stdout, payload);
}

export async function handleStream(opts: {
  ctx: AdapterContext;
  handlers: StreamHandlers;
  stdin?: NodeJS.ReadableStream;
  stdout?: NodeJS.WriteStream;
  validate?: boolean;
}): Promise<void> {
  const stdin = opts.stdin ?? process.stdin;
  const stdout = opts.stdout ?? process.stdout;
  const validate = opts.validate ?? true;
  const ctx = opts.ctx;
  const handlers = opts.handlers;

  const dispatchError = async (log: AdapterLogger, message: string) => {
    log.error("%s", message);
    emitStreamStatus({ type: "delivery_error", error: message }, stdout, validate);
  };

  for await (const raw of readJSONLines(stdin)) {
    if (ctx.signal.aborted) {
      ctx.log.info("stream shutting down (signal aborted)");
      return;
    }

    let event: StreamEvent;
    try {
      event = validate ? StreamEventSchema.parse(raw) : (raw as StreamEvent);
    } catch (err) {
      await dispatchError(ctx.log, `stream: failed to parse event: ${errorToString(err)}`);
      continue;
    }

    try {
      switch (event.type) {
        case "stream_start":
          if (handlers.onStart) {
            await handlers.onStart(ctx, event);
          }
          break;
        case "token":
          if (handlers.onToken) {
            await handlers.onToken(ctx, event);
          }
          break;
        case "tool_status":
          if (handlers.onToolStatus) {
            await handlers.onToolStatus(ctx, event);
          }
          break;
        case "reasoning":
          if (handlers.onReasoning) {
            await handlers.onReasoning(ctx, event);
          }
          break;
        case "stream_end":
          if (handlers.onEnd) {
            await handlers.onEnd(ctx, event);
          }
          break;
        case "stream_error":
          if (handlers.onError) {
            await handlers.onError(ctx, event);
          }
          break;
        default:
          ctx.log.debug("stream: unknown event type: %s", (event as { type: string }).type);
      }
    } catch (err) {
      await dispatchError(ctx.log, `stream handler error for ${event.type}: ${errorToString(err)}`);
    }
  }
}

function errorToString(err: unknown): string {
  return err instanceof Error ? err.message : String(err);
}

